File: rinthel_tui/tui/effects/tagline.py
"""Tagline rotativo — reusa glitch_frames() de flicker.py para la transición
entre frases. Reusable por cualquier screen con título (no solo el menú)."""

import random
from typing import Sequence

from textual.timer import Timer
from textual.widgets import Static

from rinthel_tui.tui.effects.flicker import glitch_frames


class RotatingTagline(Static):
    """Cicla `lines` cada `interval_s`, revelando cada frase nueva con una
    transición glitch de `transition_frames` pasos — mismo algoritmo que
    GlitchLabel usa en su reveal inicial, aplicado en loop."""

    def __init__(
        self,
        lines: Sequence[str],
        *,
        interval_s: float = 6.0,
        shuffle: bool = True,
        transition_frames: int = 6,
        frame_ms: int = 40,
        **kwargs,
    ) -> None:
        super().__init__(**kwargs)
        self._lines = list(lines)
        if shuffle:
            random.shuffle(self._lines)
        self._interval_s = interval_s
        self._transition_frames = transition_frames
        self._frame_ms = frame_ms
        self._index = 0
        self._frames: list[str] = []
        self._frame_pos = 0
        self._frame_timer: Timer | None = None

    def on_mount(self) -> None:
        self.update(self._lines[0])
        self.set_interval(self._interval_s, self._start_transition)

    def _start_transition(self) -> None:
        self._index = (self._index + 1) % len(self._lines)
        self._frames = list(glitch_frames(self._lines[self._index], self._transition_frames))
        self._frame_pos = 0
        self.update(self._frames[0])
        self._frame_timer = self.set_interval(self._frame_ms / 1000, self._advance_frame)

    def _advance_frame(self) -> None:
        self._frame_pos += 1
        if self._frame_pos >= len(self._frames):
            if self._frame_timer is not None:
                self._frame_timer.stop()
            return
        self.update(self._frames[self._frame_pos])

File: rinthel-client/src/effects/tagline.rs
//! Puerto de `tui/effects/tagline.py` — `RotatingTagline`: cicla frases,
//! revelando cada una con la misma transición glitch que `GlitchLabel`
//! (`flicker::GlitchReveal`) reusada acá, igual que en Python
//! (`from ...flicker import glitch_frames`).

use std::time::{Duration, Instant};

use rand::seq::SliceRandom;

use super::flicker::GlitchReveal;

pub struct RotatingTagline {
    lines: Vec<String>,
    index: usize,
    interval: Duration,
    transition_frames: usize,
    frame_ms: u64,
    next_rotation: Instant,
    transition: Option<GlitchReveal>,
}

impl RotatingTagline {
    pub fn new(lines: &[&str], interval_s: f64, transition_frames: usize, frame_ms: u64) -> Self {
        let mut lines: Vec<String> = lines.iter().map(|s| (*s).to_string()).collect();
        lines.shuffle(&mut rand::thread_rng());
        let interval = Duration::from_secs_f64(interval_s);
        RotatingTagline {
            lines,
            index: 0,
            interval,
            transition_frames,
            frame_ms,
            next_rotation: Instant::now() + interval,
            transition: None,
        }
    }

    /// Llamado cada vuelta del loop principal mientras el menú está activo
    /// — reemplaza los dos `set_interval` de Python (rotación + frames de
    /// transición) por un chequeo de reloj de pared.
    pub fn advance(&mut self, now: Instant) {
        if let Some(t) = &self.transition {
            if t.is_done(now) {
                self.transition = None;
            }
        }
        if self.transition.is_none() && now >= self.next_rotation && self.lines.len() > 1 {
            self.index = (self.index + 1) % self.lines.len();
            self.transition = Some(GlitchReveal::start(&self.lines[self.index], self.transition_frames, self.frame_ms));
            self.next_rotation = now + self.interval;
        }
    }

    pub fn text(&self, now: Instant) -> &str {
        match &self.transition {
            Some(t) => t.text(now),
            None => &self.lines[self.index],
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn starts_on_first_line_no_transition() {
        let tagline = RotatingTagline::new(&["uno", "dos"], 6.0, 6, 40);
        let now = Instant::now();
        assert!(tagline.lines.contains(&tagline.text(now).to_string()));
    }

    #[test]
    fn rotates_after_interval_elapses() {
        let mut tagline = RotatingTagline::new(&["uno", "dos", "tres"], 0.01, 4, 5);
        let first = tagline.index;
        std::thread::sleep(Duration::from_millis(15));
        tagline.advance(Instant::now());
        assert_ne!(tagline.index, first);
    }
}
